src/merkle_tree/incremental_merkle_tree.rs

```diff
@@ -0,0 +1,277 @@
+                    new_auth_path.push(old_auth_path[x].clone());
+
+                    // update the new current node (this is needed to compute the root)
+                    let (new_left, new_right) = if is_left_child(new_index) {
+                        (new_current_node, hash_of_empty_node.clone())
+                    } else {
+                        (old_auth_path[x].clone(), new_current_node)
+                    };
+                    new_current_node = P::TwoToOneHash::compress(
+                        &self.two_to_one_hash_param,
+                        new_left,
+                        new_right,
+                    )?;
+                } else {
+                    // this means old path and new path haven't been merged,
+                    // as a reulst, need to update both the new_current_node and new_current_node
+                    let auth_node = if is_left_child(new_index) {
+                        hash_of_empty_node.clone()
+                    } else {
+                        old_current_node.clone()
+                    };
+                    new_auth_path.push(auth_node);
+
+                    // update both old_current_node and new_current_node
+                    // update new_current_node
+                    let (new_left, new_right) = if is_left_child(new_index) {
+                        (new_current_node.clone(), hash_of_empty_node.clone())
+                    } else {
+                        (old_current_node.clone(), new_current_node)
+                    };
+                    new_current_node = P::TwoToOneHash::compress(
+                        &self.two_to_one_hash_param,
+                        new_left,
+                        new_right,
+                    )?;
+
+                    // We only need to update the old_current_node bottom up when it is right child
+                    if !is_left_child(old_index) {
+                        old_current_node = P::TwoToOneHash::compress(
+                            &self.two_to_one_hash_param,
+                            old_auth_path[x].clone(),
+                            old_current_node,
+                        )?;
+                    }
+                }
+            }
+
+            // reverse new_auth_path to top down
+            new_auth_path.reverse();
+            let path = Path {
+                leaf_index: self.next_available().unwrap(),
+                auth_path: new_auth_path,
+                leaf_sibling_hash: leaf_sibling,
+            };
+            Ok((path, new_current_node))
+        }
+    }
+
+    /// the proof of the current item
+    pub fn current_proof(&self) -> Path<P> {
+        self.current_path.clone()
+    }
+
+    /// root of IMT
+    pub fn root(&self) -> P::InnerDigest {
+        self.root.clone()
+    }
+}
+
+/// Return true iff the given index on its current level represents a left child
+#[inline]
+fn is_left_child(index_on_level: usize) -> bool {
+    index_on_level % 2 == 0
```

Comment:
Compiler will optimize this to ``index_on_level & 1 == 0`` right ?

Reply:
yeah